[Bug] [savepoints] CDC SOURCE The operator graph generated by each submission task is inconsistent, resulting in the failure to restore the old and new serialization of the savepoints.

### Search before asking

- [X] I had searched in the [issues](https://github.com/DataLinkDC/dinky/issues?q=is%3Aissue) and found no similar issues.


### What happened

采用的是dlink本身的提交流程，提交SQL未发生变化，算子不应该出现问题才对
早期测试：
两张表同步任务这样操作是可以正常恢复
使用3张表提交记录保存点，再恢复不行，尝试等待10分钟后记录保存点并取消任务，依然恢复失败
**中间没有多加或者少表步骤，只是测试的表数量不一样，但恢复的保存点记录操作表是一样的**

**后续测试：
2张表不一定百分比成功，除非恢复任务生成的算子图跟上次提交的算子图一致才能成功，否则失败
结论：若多表的情况下， 执行任务和恢复任务生成的算子图顺序的排列组合越来越多，每次生成算子图表的顺序一致的概率会大大降低，导致按保存点恢复任务成功的概率很少**

confProperties["execution.savepoint.path"] = “file:/tmp/flink/savepoints/savepoint-7a7365-ef55ca3e1604” 恢复任务，提示：
`Caused by: org.apache.flink.runtime.state.BackendBuildingException: Caught unexpected exception.
	at org.apache.flink.contrib.streaming.state.RocksDBKeyedStateBackendBuilder.build(RocksDBKeyedStateBackendBuilder.java:409)
	at org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend.createKeyedStateBackend(EmbeddedRocksDBStateBackend.java:505)
	at org.apache.flink.contrib.streaming.state.EmbeddedRocksDBStateBackend.createKeyedStateBackend(EmbeddedRocksDBStateBackend.java:98)
	at org.apache.flink.streaming.api.operators.StreamTaskStateInitializerImpl.lambda$keyedStatedBackend$1(StreamTaskStateInitializerImpl.java:336)
	at org.apache.flink.streaming.api.operators.BackendRestorerProcedure.attemptCreateAndRestore(BackendRestorerProcedure.java:168)
	at org.apache.flink.streaming.api.operators.BackendRestorerProcedure.createAndRestore(BackendRestorerProcedure.java:135)
	... 13 more
Caused by: org.apache.flink.util.StateMigrationException: The new key serializer (org.apache.flink.table.runtime.typeutils.RowDataSerializer@58a0d672) must be compatible with the previous key serializer (org.apache.flink.table.runtime.typeutils.RowDataSerializer@24d).
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.readMetaData(RocksDBIncrementalRestoreOperation.java:549)
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.readMetaData(RocksDBIncrementalRestoreOperation.java:516)
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.restoreFromLocalState(RocksDBIncrementalRestoreOperation.java:232)
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.restoreFromRemoteState(RocksDBIncrementalRestoreOperation.java:222)
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.restoreWithoutRescaling(RocksDBIncrementalRestoreOperation.java:189)
	at org.apache.flink.contrib.streaming.state.restore.RocksDBIncrementalRestoreOperation.restore(RocksDBIncrementalRestoreOperation.java:169)
	at org.apache.flink.contrib.streaming.state.RocksDBKeyedStateBackendBuilder.build(RocksDBKeyedStateBackendBuilder.java:329)
	... 18 more
2024-03-22 11:11:17,318 INFO  org.apache.flink.runtime.dispatcher.StandaloneDispatcher     [] - Job c3664a9c7786cfc963452fec76db7928 has been registered for cleanup in the JobResultStore after reaching a terminal state.`


submit FlinkSQL：
EXECUTE CDCSOURCE jira.TY002-JIRA-SDI2_2TEST_ODS_JIRA.TY999-DW WITH (
   'connector' = 'mysql-cdc',
   'hostname' = '192.168.10.171',
   'port' = '3336',
   'username' = 'read_jira',
   'password' = 'xxxxxx',
   'database-name' = 'jira',
   'parallelism' = '1',
   'checkpoint' = '300000',
   'scan.startup.mode' = 'latest-offset',
   'table-name' = 'jira\.jiraaction,jira\.fileattachment,jira\.issuestatus,jira\.jiraissue,jira\.label,jira\.issuetype,jira\.project',
   'sink.table.prefix' = 'ODS_',
   'sink.table-name' ='SDI2_2TEST_ODS_JIRA.${tableName}',
   'sink.connector' = 'jdbc',
   'sink.url' = 'jdbc:oracle:thin:@//192.168.10.161:1521/dev',
   'sink.username' = 'zt_2_2_test',
   'sink.password' = 'xxxxxx')


### What you expected to happen

本地测试使用LocalStreamExecutor方式提交并记录保存点文件，再恢复没有问题，而远程RemoteStreamExecutor对象记录的保存点失败，按理说本地可以说明算子本身没有发生变化，不应该失败才对，辛苦作者帮忙看下或者有没有遇到相似的问题

### How to reproduce

以当前时间提交任务，然后调用flink restAPI记录保存点并取消任务，这时候存储数据状态的rocksb文件会被自动dispose掉，保存点仅.metadata文件，再配置：confProperties["execution.savepoint.path"] = executionSavepointPath 即可复现

### Anything else

_No response_

### Version

0.7.4

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
@@ -51,6 +51,7 @@ import org.apache.flink.table.types.logical.FloatType;
 import org.apache.flink.table.types.logical.IntType;
 import org.apache.flink.table.types.logical.LogicalType;
 import org.apache.flink.table.types.logical.SmallIntType;
+import org.apache.flink.table.types.logical.TimeType;
 import org.apache.flink.table.types.logical.TimestampType;
 import org.apache.flink.table.types.logical.TinyIntType;
 import org.apache.flink.table.types.logical.VarBinaryType;
@@ -64,6 +65,7 @@ import java.time.Instant;
 import java.time.ZoneId;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
@@ -102,6 +104,7 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
         typeConverterList = Lists.newArrayList(
                 this::convertVarCharType,
                 this::convertDateType,
+                this::convertTimeType,
                 this::convertVarBinaryType,
                 this::convertBigIntType,
                 this::convertFloatType,
@@ -271,15 +274,30 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
         config.getSink().remove("timezone");
         if (Asserts.isNotNullString(timeZone)) {
             sinkTimeZone = ZoneId.of(timeZone);
+            logger.info("Sink timezone is {}", sinkTimeZone);
         }
 
         final List<Schema> schemaList = config.getSchemaList();
+        if (Asserts.isNullCollection(schemaList)) {
+            logger.warn("Schema list is empty, please check your configuration and try again.");
+            return dataStreamSource;
+        }
+
         final String schemaFieldName = config.getSchemaFieldName();
 
         if (Asserts.isNotNullCollection(schemaList)) {
             SingleOutputStreamOperator<Map> mapOperator = deserialize(dataStreamSource);
             for (Schema schema : schemaList) {
-                for (Table table : schema.getTables()) {
+                if (Asserts.isNullCollection(schema.getTables())) {
+                    // if schema tables is empty, throw exception
+                    throw new IllegalArgumentException(
+                            "Schema tables is empty, please check your configuration or check your database permission and try again.");
+                }
+                // if schema tables is not empty, sort by table name
+                List<Table> tableList = schema.getTables().stream()
+                        .sorted(Comparator.comparing(Table::getName))
+                        .collect(Collectors.toList());
+                for (Table table : tableList) {
                     SingleOutputStreamOperator<Map> filterOperator = shunt(mapOperator, table, schemaFieldName);
 
                     List<String> columnNameList = new ArrayList<>();
@@ -338,6 +356,11 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
             case DATE:
             case LOCAL_DATE:
                 return new DateType();
+            case TIME:
+            case LOCALTIME:
+                return column.getPrecision() != null
+                        ? new TimeType(column.isNullable(), column.getPrecision())
+                        : new TimeType();
             case LOCAL_DATETIME:
             case TIMESTAMP:
                 if (column.getLength() != null) {
@@ -447,6 +470,16 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
             return Optional.of(StringData.fromString(Instant.ofEpochMilli((long) target)
                     .atZone(ZoneId.systemDefault())
                     .toLocalDate()
+                    .toString()));
+        }
+        return Optional.empty();
+    }
+
+    protected Optional<Object> convertTimeType(Object target, LogicalType logicalType) {
+        if (logicalType instanceof TimeType) {
+            return Optional.of(StringData.fromString(Instant.ofEpochMilli((long) target)
+                    .atZone(ZoneId.systemDefault())
+                    .toLocalTime()
                     .toString()));
         }
         return Optional.empty();

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
@@ -324,8 +324,6 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
 
     public LogicalType getLogicalType(Column column) {
         switch (column.getJavaType()) {
-            case STRING:
-                return new VarCharType();
             case BOOLEAN:
             case JAVA_LANG_BOOLEAN:
                 return new BooleanType();
@@ -358,9 +356,11 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
                 return new DateType();
             case TIME:
             case LOCALTIME:
-                return column.getPrecision() != null
-                        ? new TimeType(column.isNullable(), column.getPrecision())
-                        : new TimeType();
+                if (column.getPrecision() != null) {
+                    return new TimeType(column.isNullable(), column.getPrecision());
+                } else {
+                    return new TimeType();
+                }
             case LOCAL_DATETIME:
             case TIMESTAMP:
                 if (column.getLength() != null) {
@@ -370,6 +370,7 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
                 }
             case BYTES:
                 return new VarBinaryType(Integer.MAX_VALUE);
+            case STRING:
             default:
                 return new VarCharType();
         }

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
@@ -355,7 +355,6 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
             case LOCAL_DATE:
                 return new DateType();
             case TIME:
-            case LOCALTIME:
                 if (column.getPrecision() != null) {
                     return new TimeType(column.isNullable(), column.getPrecision());
                 } else {

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/AbstractSinkBuilder.java
@@ -354,12 +354,6 @@ public abstract class AbstractSinkBuilder implements SinkBuilder {
             case DATE:
             case LOCAL_DATE:
                 return new DateType();
-            case TIME:
-                if (column.getPrecision() != null) {
-                    return new TimeType(column.isNullable(), column.getPrecision());
-                } else {
-                    return new TimeType();
-                }
             case LOCAL_DATETIME:
             case TIMESTAMP:
                 if (column.getLength() != null) {

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/kafka/KafkaSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/kafka/KafkaSinkBuilder.java
@@ -63,11 +63,12 @@ import org.apache.flink.util.Collector;
 import org.apache.flink.util.OutputTag;
 
 import java.io.Serializable;
-import java.util.HashMap;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.stream.Collectors;
 
 /**
  * MysqlCDCBuilder
@@ -128,17 +129,27 @@ public class KafkaSinkBuilder extends AbstractSinkBuilder implements Serializabl
             KafkaSink<String> kafkaSink = kafkaSinkBuilder.build();
             dataStreamSource.sinkTo(kafkaSink);
         } else {
-            Map<Table, OutputTag<String>> tagMap = new HashMap<>();
-            Map<String, Table> tableMap = new HashMap<>();
+            Map<Table, OutputTag<String>> tagMap = new LinkedHashMap<>();
+            Map<String, Table> tableMap = new LinkedHashMap<>();
             ObjectMapper objectMapper = new ObjectMapper();
             SingleOutputStreamOperator<Map> mapOperator = dataStreamSource
                     .map(x -> objectMapper.readValue(x, Map.class))
                     .returns(Map.class);
             final List<Schema> schemaList = config.getSchemaList();
+
             final String schemaFieldName = config.getSchemaFieldName();
             if (Asserts.isNotNullCollection(schemaList)) {
                 for (Schema schema : schemaList) {
-                    for (Table table : schema.getTables()) {
+                    if (Asserts.isNullCollection(schema.getTables())) {
+                        // if schema tables is empty, throw exception
+                        throw new IllegalArgumentException(
+                                "Schema tables is empty, please check your configuration or check your database permission and try again.");
+                    }
+                    // if schema tables is not empty, sort by table name
+                    List<Table> tableList = schema.getTables().stream()
+                            .sorted(Comparator.comparing(Table::getName))
+                            .collect(Collectors.toList());
+                    for (Table table : tableList) {
                         String sinkTableName = getSinkTableName(table);
                         OutputTag<String> outputTag = new OutputTag<String>(sinkTableName) {};
                         tagMap.put(table, outputTag);

--- a/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/sql/AbstractSqlSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-core/src/main/java/org/dinky/cdc/sql/AbstractSqlSinkBuilder.java
@@ -51,7 +51,7 @@ import java.io.Serializable;
 import java.math.BigDecimal;
 import java.time.ZoneId;
 import java.util.ArrayList;
-import java.util.HashMap;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
@@ -207,6 +207,7 @@ public abstract class AbstractSqlSinkBuilder extends AbstractSinkBuilder impleme
         config.getSink().remove("timezone");
         if (Asserts.isNotNullString(timeZone)) {
             sinkTimeZone = ZoneId.of(timeZone);
+            logger.info("Sink timezone is {}", sinkTimeZone);
         }
 
         final List<Schema> schemaList = config.getSchemaList();
@@ -218,10 +219,19 @@ public abstract class AbstractSqlSinkBuilder extends AbstractSinkBuilder impleme
 
         logger.info("Build deserialize successful...");
 
-        Map<Table, OutputTag<Map>> tagMap = new HashMap<>();
-        Map<String, Table> tableMap = new HashMap<>();
+        Map<Table, OutputTag<Map>> tagMap = new LinkedHashMap<>();
+        Map<String, Table> tableMap = new LinkedHashMap<>();
         for (Schema schema : schemaList) {
-            for (Table table : schema.getTables()) {
+            if (Asserts.isNullCollection(schema.getTables())) {
+                // if schema tables is empty, throw exception
+                throw new IllegalArgumentException(
+                        "Schema tables is empty, please check your configuration or check your database permission and try again.");
+            }
+            // if schema tables is not empty, sort by table name
+            List<Table> tableList = schema.getTables().stream()
+                    .sorted(Comparator.comparing(Table::getName))
+                    .collect(Collectors.toList());
+            for (Table table : tableList) {
                 String sinkTableName = getSinkTableName(table);
                 OutputTag<Map> outputTag = new OutputTag<Map>(sinkTableName) {};
                 tagMap.put(table, outputTag);

--- a/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/doris/DorisSchemaEvolutionSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/doris/DorisSchemaEvolutionSinkBuilder.java
@@ -41,12 +41,12 @@ import org.apache.flink.util.Collector;
 import org.apache.flink.util.OutputTag;
 
 import java.io.Serializable;
-import java.util.HashMap;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
-import java.util.UUID;
+import java.util.stream.Collectors;
 
 public class DorisSchemaEvolutionSinkBuilder extends AbstractSinkBuilder implements Serializable {
 
@@ -92,9 +92,19 @@ public class DorisSchemaEvolutionSinkBuilder extends AbstractSinkBuilder impleme
                 dataStreamSource.map(x -> objectMapper.readValue(x, Map.class)).returns(Map.class);
         final String schemaFieldName = config.getSchemaFieldName();
 
-        Map<Table, OutputTag<String>> tagMap = new HashMap<>();
-        Map<String, Table> tableMap = new HashMap<>();
+        Map<Table, OutputTag<String>> tagMap = new LinkedHashMap<>();
+        Map<String, Table> tableMap = new LinkedHashMap<>();
         for (Schema schema : schemaList) {
+            if (Asserts.isNullCollection(schema.getTables())) {
+                // if schema tables is empty, throw exception
+                throw new IllegalArgumentException(
+                        "Schema tables is empty, please check your configuration or check your database permission and try again.");
+            }
+            // if schema tables is not empty, sort by table name
+            List<Table> tableList = schema.getTables().stream()
+                    .sorted(Comparator.comparing(Table::getName))
+                    .collect(Collectors.toList());
+
             for (Table table : schema.getTables()) {
                 OutputTag<String> outputTag = new OutputTag<String>(getSinkTableName(table)) {};
                 tagMap.put(table, outputTag);
@@ -150,7 +160,8 @@ public class DorisSchemaEvolutionSinkBuilder extends AbstractSinkBuilder impleme
                         getSinkSchemaName(table),
                         getSinkTableName(table)));
             } else {
-                // flink-cdc-pipeline-connector-doris 3.0.0 以上版本内部已经拼接了 SchemaName + SinkTableName，并且约定 TableLabel 正则表达式如下 --> regex: ^[-_A-Za-z0-9]{1,128}$
+                // flink-cdc-pipeline-connector-doris 3.0.0 以上版本内部已经拼接了 SchemaName + SinkTableName，并且约定 TableLabel
+                // 正则表达式如下 --> regex: ^[-_A-Za-z0-9]{1,128}$
                 executionBuilder.setLabelPrefix("dinky");
             }
 
@@ -160,11 +171,13 @@ public class DorisSchemaEvolutionSinkBuilder extends AbstractSinkBuilder impleme
 
             executionBuilder.setStreamLoadProp(properties).setDeletable(true);
 
-            JsonDebeziumSchemaSerializer.Builder jsonDebeziumSchemaSerializerBuilder = JsonDebeziumSchemaSerializer.builder();
+            JsonDebeziumSchemaSerializer.Builder jsonDebeziumSchemaSerializerBuilder =
+                    JsonDebeziumSchemaSerializer.builder();
 
             // use new schema change
             if (sink.containsKey(DorisSinkOptions.SINK_USE_NEW_SCHEMA_CHANGE.key())) {
-                jsonDebeziumSchemaSerializerBuilder.setNewSchemaChange(Boolean.valueOf(sink.get(DorisSinkOptions.SINK_USE_NEW_SCHEMA_CHANGE.key())));
+                jsonDebeziumSchemaSerializerBuilder.setNewSchemaChange(
+                        Boolean.valueOf(sink.get(DorisSinkOptions.SINK_USE_NEW_SCHEMA_CHANGE.key())));
             }
 
             DorisSink.Builder<String> builder = DorisSink.builder();

--- a/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/kafka/KafkaSinkBuilder.java
+++ b/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/kafka/KafkaSinkBuilder.java
@@ -41,11 +41,12 @@ import org.apache.flink.util.Collector;
 import org.apache.flink.util.OutputTag;
 
 import java.io.Serializable;
-import java.util.HashMap;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
+import java.util.stream.Collectors;
 
 public class KafkaSinkBuilder extends AbstractSinkBuilder implements Serializable {
 
@@ -97,17 +98,27 @@ public class KafkaSinkBuilder extends AbstractSinkBuilder implements Serializabl
             KafkaSink<String> kafkaSink = kafkaSinkBuilder.build();
             dataStreamSource.sinkTo(kafkaSink);
         } else {
-            Map<Table, OutputTag<String>> tagMap = new HashMap<>();
-            Map<String, Table> tableMap = new HashMap<>();
+            Map<Table, OutputTag<String>> tagMap = new LinkedHashMap<>();
+            Map<String, Table> tableMap = new LinkedHashMap<>();
             ObjectMapper objectMapper = new ObjectMapper();
             SingleOutputStreamOperator<Map> mapOperator = dataStreamSource
                     .map(x -> objectMapper.readValue(x, Map.class))
                     .returns(Map.class);
             final List<Schema> schemaList = config.getSchemaList();
+
             final String schemaFieldName = config.getSchemaFieldName();
             if (Asserts.isNotNullCollection(schemaList)) {
                 for (Schema schema : schemaList) {
-                    for (Table table : schema.getTables()) {
+                    if (Asserts.isNullCollection(schema.getTables())) {
+                        // if schema tables is empty, throw exception
+                        throw new IllegalArgumentException(
+                                "Schema tables is empty, please check your configuration or check your database permission and try again.");
+                    }
+                    // if schema tables is not empty, sort by table name
+                    List<Table> tableList = schema.getTables().stream()
+                            .sorted(Comparator.comparing(Table::getName))
+                            .collect(Collectors.toList());
+                    for (Table table : tableList) {
                         String sinkTableName = getSinkTableName(table);
                         OutputTag<String> outputTag = new OutputTag<String>(sinkTableName) {};
                         tagMap.put(table, outputTag);

--- a/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/kafka/KafkaSinkJsonBuilder.java
+++ b/dinky-cdc/dinky-cdc-plus/src/main/java/org/dinky/cdc/kafka/KafkaSinkJsonBuilder.java
@@ -43,10 +43,12 @@ import org.apache.flink.util.Collector;
 import java.io.Serializable;
 import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
+import java.util.Comparator;
 import java.util.LinkedHashMap;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.SerializationFeature;
@@ -91,7 +93,16 @@ public class KafkaSinkJsonBuilder extends AbstractSinkBuilder implements Seriali
             }
 
             for (Schema schema : schemaList) {
-                for (Table table : schema.getTables()) {
+                if (Asserts.isNullCollection(schema.getTables())) {
+                    // if schema tables is empty, throw exception
+                    throw new IllegalArgumentException(
+                            "Schema tables is empty, please check your configuration or check your database permission and try again.");
+                }
+                // if schema tables is not empty, sort by table name
+                List<Table> tableList = schema.getTables().stream()
+                        .sorted(Comparator.comparing(Table::getName))
+                        .collect(Collectors.toList());
+                for (Table table : tableList) {
                     final String tableName = table.getName();
                     final String schemaName = table.getSchema();
                     SingleOutputStreamOperator<Map> filterOperator = mapOperator.filter((FilterFunction<Map>) value -> {

--- a/dinky-common/src/main/java/org/dinky/data/model/Schema.java
+++ b/dinky-common/src/main/java/org/dinky/data/model/Schema.java
@@ -20,7 +20,7 @@
 package org.dinky.data.model;
 
 import java.io.Serializable;
-import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.List;
 
 import lombok.Getter;
@@ -38,11 +38,11 @@ public class Schema implements Serializable, Comparable<Schema> {
     private static final long serialVersionUID = 4278304357661271040L;
 
     private String name;
-    private List<Table> tables = new ArrayList<>();
-    private List<String> views = new ArrayList<>();
-    private List<String> functions = new ArrayList<>();
-    private List<String> userFunctions = new ArrayList<>();
-    private List<String> modules = new ArrayList<>();
+    private List<Table> tables = new LinkedList<>();
+    private List<String> views = new LinkedList<>();
+    private List<String> functions = new LinkedList<>();
+    private List<String> userFunctions = new LinkedList<>();
+    private List<String> modules = new LinkedList<>();
 
     /** 需要保留一个空构造方法，否则序列化有问题 */
     public Schema() {}

--- a/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
+++ b/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
@@ -212,7 +212,8 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
             driver.createSchema(schema);
         }
         sink.put(FlinkCDCConfig.SINK_DB, schema);
-        // todo: There is a bug that can cause the problem of URL duplicate concatenation of schema, for example: jdbc: mysql://localhost:3306/test?useSSL=false/test -1
+        // todo: There is a bug that can cause the problem of URL duplicate concatenation of schema, for example: jdbc:
+        // mysql://localhost:3306/test?useSSL=false/test -1
         if (!url.contains(schema)) {
             sink.put("url", url + "/" + schema);
         }

--- a/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
+++ b/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
@@ -39,8 +39,9 @@ import org.apache.flink.streaming.api.datastream.DataStreamSource;
 import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
 import org.apache.flink.table.api.TableResult;
 
-import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
@@ -81,16 +82,19 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
             Map<String, Map<String, String>> allConfigMap = cdcBuilder.parseMetaDataConfigs();
             config.setSchemaFieldName(cdcBuilder.getSchemaFieldName());
             SinkBuilder sinkBuilder = SinkBuilderFactory.buildSinkBuilder(config);
-            List<Schema> schemaList = new ArrayList<>();
             final List<String> schemaNameList = cdcBuilder.getSchemaList();
             final List<String> tableRegList = cdcBuilder.getTableList();
-            final List<String> schemaTableNameList = new ArrayList<>();
+
+            final List<Schema> schemaList = new LinkedList<>();
+            final List<String> schemaTableNameList = new LinkedList<>();
+            // Scenario of dividing databases and tables
             if (SplitUtil.isEnabled(cdcSource.getSplit())) {
+                logger.info("Split table or database mode is enabled...");
                 Map<String, String> confMap = cdcBuilder.parseMetaDataConfig();
                 Driver driver =
                         Driver.buildWithOutPool(confMap.get("name"), confMap.get("type"), JsonUtils.toMap(confMap));
 
-                // 这直接传正则过去
+                // This is passed directly to the regularization process
                 schemaTableNameList.addAll(tableRegList.stream()
                         .map(x -> x.replaceFirst("\\\\.", "."))
                         .collect(Collectors.toList()));
@@ -100,12 +104,17 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
                 Set<Table> tables = driver.getSplitTables(tableRegList, cdcSource.getSplit());
 
                 for (Table table : tables) {
+                    // Filter out views
+                    if (Asserts.isEquals(table.getType(), "VIEW")) {
+                        continue;
+                    }
                     String schemaName = table.getSchema();
                     Schema schema = Schema.build(schemaName);
                     schema.setTables(Collections.singletonList(table));
-                    // 分库分表所有表结构都是一样的，取出列表中第一个表名即可
+                    // The structure of all tables in a database or table is the same, just take out the first table
+                    // name from the list
                     String schemaTableName = table.getSchemaTableNameList().get(0);
-                    // 真实的表名
+                    // Real Table Name
                     String realSchemaName = schemaTableName.split("\\.")[0];
                     String tableName = schemaTableName.split("\\.")[1];
                     table.setColumns(driver.listColumnsSortByPK(realSchemaName, tableName));
@@ -167,7 +176,9 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
                 logger.info("{}: {}", i + 1, schemaTableNameList.get(i));
             }
             config.setSchemaTableNameList(schemaTableNameList);
-            config.setSchemaList(schemaList);
+            config.setSchemaList(schemaList.stream()
+                    .sorted(Comparator.comparing(Schema::getName))
+                    .collect(Collectors.toList()));
             StreamExecutionEnvironment streamExecutionEnvironment = executor.getStreamExecutionEnvironment();
             if (Asserts.isNotNull(config.getParallelism())) {
                 streamExecutionEnvironment.setParallelism(config.getParallelism());
@@ -201,6 +212,7 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
             driver.createSchema(schema);
         }
         sink.put(FlinkCDCConfig.SINK_DB, schema);
+        // todo: 存在 bug, 会导致 url 重复拼接 schema 的问题 , 例如: jdbc:mysql://localhost:3306/test?useSSL=false/test-1
         if (!url.contains(schema)) {
             sink.put("url", url + "/" + schema);
         }

--- a/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
+++ b/dinky-core/src/main/java/org/dinky/trans/ddl/CreateCDCSourceOperation.java
@@ -212,7 +212,7 @@ public class CreateCDCSourceOperation extends AbstractOperation implements Opera
             driver.createSchema(schema);
         }
         sink.put(FlinkCDCConfig.SINK_DB, schema);
-        // todo: 存在 bug, 会导致 url 重复拼接 schema 的问题 , 例如: jdbc:mysql://localhost:3306/test?useSSL=false/test-1
+        // todo: There is a bug that can cause the problem of URL duplicate concatenation of schema, for example: jdbc: mysql://localhost:3306/test?useSSL=false/test -1
         if (!url.contains(schema)) {
             sink.put("url", url + "/" + schema);
         }

